### PR TITLE
46233002_Siamese_Triplet_Net

### PR DESCRIPTION
**Pull request by eva0408**
Pull request: Siamese-network-based Classifier on ADNI Brain Dataset created by student 46233002.

Summary:
This pull request aims to add a Siamese network-based classifier on ADNI Brain Dataset. It has an architecture of an encoder trained using triplet loss and a classifier trained using cross-entropy loss. It uses several data augmentation and a miner for generating triplet combinations to enhance performance. The model has achieved an accuracy of 72% on ADNI Brain Dataset.

Files:
- 'README.md' contains the complete information and analysis of this pull request and the model.
- 'dataset.py' contains functions required for pre-processing the dataset. 
- 'modules.py' contains the full implementation of the model
- 'predict.py' is the file used to make predictions on test data using a trained model.
- 'train.py' is the file used to train the model.
- 'utils.py' contains some helper functions required for the training, validating, and testing of the model. 

Student name: Eva Chang
Student ID: 46233002

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Good
- Clear flow of information
- Algorithm, data handling, and training details explained well
- not enough results/plots included

Commit messages: Good

Code: 
- looking at the plot, triplet loss doesn't look very right, could be an issue with the code
- good try with mining strategies, but doesn't seem to have helped the network
- good design
- sufficiently commented

Functionality/Performance: 
- classifier only achieves 72% for test data
- triplet network seem to be performing poorly too

General comments:
- **PR is in the main branch**
- solves the problem apporpriately with triplet network followed by a classifier, although performance is very poor

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem, low performance/triplet loss issue -1
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in InCorrect Branch) -2
Feedback required, please amend the branch -2
Request Description minimal -1

**Comment by shakes76** [comment]
Feedback marks possible +2 if the requested changes are made (see above).

**Comment by wangzhaomxy** [comment]
Feedback attempt and feedback marks granted +2.
